### ISSUE TITLE
Consistency between write/read operation on the same file

### ISSUE THREAD

**Issue by jbreitbart** [axboe/liburing#329]
I am currently debugging various test cases that use liburing. The tests are not failing every time, but it looks like they are randomly reading wrong data. As far as I can tell all affected tests follow the same pattern:

* write data to the file using liburing
* wait for all requests **(late edit)**
* close the file descriptor and call `io_uring_queue_exit`
* open the same file again
* read data using `io_uring_prep_readv`
* wait for the data via `io_uring_wait_cqe_nr`
* wrong data is available

Is waiting + calling `io_uring_queue_exit` enough to make the data available for future reads?

I have created a stand alone testcase here: https://gist.github.com/jbreitbart/c9a2f03f08de6abb8d161a708238bf6f It consistently fails on my system. The content of the file written by the test is fine after the test reported the error.

I am using Ubuntu 18.04 with liburing-2.0. I have updated the kernel to 5.12.0-051200rc7-generic, but I saw the same behavior with older kernel versions as well.

**Comment by isilence** [comment]
io_uring_queue_exit() doesn't guarantee that everything has been cancelled and cleaned up. It will do cancellations in the background, that also means that it may modify files and complete other operations after the function has returned. And for the same reasons used files (especially if they registered) may stay alive (i.e. not closed) until ring fully dies

**Comment by jbreitbart** [comment]
Thanks for the quick answer. That explains why my tests are currently failing. Is there a way to guarantee that data was written and can be read again? Is it guaranteed to be available after the process that wrote the data has ended?

**Comment by isilence** [comment]
> Is there a way to guarantee that data was written and can be read again?

Some of `io_uring_register(2)` operations do full quiesce, means wait for all requests to complete. And same effect you may achieve with a request marked `IOSQE_IO_DRAIN`, it won't be issue until all previous reqs have been completed, hence

```C
sqe = get_sqe();
sqe->opcode = NOP;
sqe->flags = IOSQE_IO_DRAIN;
sqe->user_data = <your_tag>;

while (cqe = wait_cqe())
    if (cqe->user_data ==  <your_tag>)
        break;
```

Drawbacks:
1. You may want to do fsync/shutdown/etc., because it only waits for requests to be killed, ring destruction (and e.g. fixed file refs putting), may take some time. So, you may also want to do `IORING_UNREGISTER_FILES`, if used. Or other `io_uring_register(2)` unregister operations.

2. It waits but not actively cancels requests, so if some hang (e.g. recv() with no incoming data), you will be stuck until something happens (e.g. signal).

> Is it guaranteed to be available after the process that wrote the data has ended?

Nope, some users don't expect requests being cancelled on exit

**Comment by isilence** [comment]
Or you may cancel some of them yourself, see `IORING_OP_POLL_REMOVE`, `IORING_OP_TIMEOUT_REMOVE`. Or more generic `IORING_OP_ASYNC_CANCEL`.

**Comment by jbreitbart** [comment]
I just realized that I did forget to mention that I already wait for every outstanding request before calling `io_uring_queue_exit`. I have edited my original post, sorry for that. I'll check your suggestions tomorrow.

**Comment by YoSTEALTH** [comment]
Clearly `io_uring_queue_init` and `io_uring_queue_exit` should be done outside of the `while` loop. Unless the goal was to break it, then maybe some kind of "LINK" or "LOCK" would need to be implemented.

**Comment by YoSTEALTH** [comment]
Another issue could be that ring `fd` is already closed, you could check if its active before closing.
```python
if ring.ring_fd:
    io_uring_queue_exit(ring)
```

**Comment by jbreitbart** [comment]
> Or you may cancel some of them yourself, see `IORING_OP_POLL_REMOVE`, `IORING_OP_TIMEOUT_REMOVE`. Or more generic `IORING_OP_ASYNC_CANCEL`.

As far as I understand there should be no outstanding operations, at least I waited on every operation I submitted to the ring. I have still added a NOP with IOSQE_IO_DRAIN + wait in my test. The results are still the same.

> Clearly `io_uring_queue_init` and `io_uring_queue_exit` should be done outside of the `while` loop. Unless the goal was to break it, then maybe some kind of "LINK" or "LOCK" would need to be implemented (which I think would be too much to ask for).

Actually this was done intentionally. I was not aware that I must call `io_uring_queue_init` and `io_uring_queue_exit` only once per process. I have changed my test, however it still fails.

> Another issue could be that ring `fd` is already closed, you could check if its active before closing.
> 
> ```python
> if ring.ring_fd:
>     io_uring_queue_exit(ring)
> ```

Thanks. I added a check to my tests, but that was apparently not the issue.


I have updated my test available here: https://gist.github.com/jbreitbart/c9a2f03f08de6abb8d161a708238bf6f

Here is what it is doing:
* call `io_uring_queue_init`
* start endless loop
* call `open(filename, O_WRONLY)`
* submit 2 writes with two calls to `io_uring_prep_writev`
* wait for 2 operation with `io_uring_wait_cqe_nr`
* add an sqe with flags==IOSQE_IO_DRAIN using `io_uring_prep_nop`
* add an sqe with `io_uring_prep_fsync`
* again wait for 2 operations
* call `close`
* call `open(filename, O_RDONLY);` to open the file again
* use `io_uring_prep_readv` to read the data submitted by the previous writes
* wait for 1 operation
* check the content of the data

The check consistently fails at element 1000. The file on disk seems to be fine, and cqe->res for the read operations is identical to the amount of data written by both writes.

**Comment by YoSTEALTH** [comment]
I don't see `io_uring_queue_exit` in your updated code.

**Comment by isilence** [comment]
btw, it allocates `new size_t[data_size]`, which is `data_size * sizeof(size_t)`, but in iovec it's just `data_size`.
Can you check that it's not mixed up?

**Comment by jbreitbart** [comment]
> I don't see `io_uring_queue_exit` in your updated code.

That would have been after the endless loop, so I did not add it. I tried your suggestion on the original code I am debugging.

> btw, it allocates `new size_t[data_size]`, which is of size `data_size * sizeof(size_t)`, but in iovec it's just `data_size`.
> Can you check that it's not mixed up?

The more people you tell about your problem, the more likely it will be embarrassing. Thank you. That code works just fine now. I have updated the gist: https://gist.github.com/jbreitbart/c9a2f03f08de6abb8d161a708238bf6f in case someone is interested. I still see strange issues with the testcases I am debugging, but that is most likely on me as well, so I am closing this.
